### PR TITLE
CRM-19029 Additional tests for merge function

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
- [CRM-19029: Additional unit tests for batch merge](https://issues.civicrm.org/jira/browse/CRM-19029)

### CODE REVIEW

**Comment by jitendrapurohit** [comment]
while executing these test on my local, I get a failure on `first_name_conflict`. Is this replicated by you as well or am I missing anything ?

    1) api_v3_JobTest::testBatchMergeWorks with data set "first_name_conflict" (array('safe', array(array('Michael', 'Jackson', 'michael@neverland.com', 'Individual', 'Student'), array('Dianna', 'Jackson', 'michael@neverland.com', 'Individual', 'Student')), 1, 0, array(array('Dianna', 'Jackson', 'michael@neverland.com', 'Individual'), array('Michael', 'Jackson', 'michael@neverland.com', 'Individual'))))
    Failed asserting that two strings are equal.
    --- Expected
    +++ Actual
    @@ @@
    -'Dianna'
    +'Michael'

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I updated it to include coverage - but I think it also gets around the error you had (which I think was not there when running as a full class)

**Comment by jitendrapurohit** [comment]
I'm executing it as a full class, and still finding this error on `first_name_conflict` after the update.

**Comment by eileenmcnaughton** [comment]
Hmm I just ran the jobTest class again and didn't get the error. I don't know what to make of your error - the first contact to be created should be 'Michael' so calling contact.get should return him first...

    $result = $this->callAPISuccess('Contact', 'get', array(
      'contact_sub_type' => 'Student',
      'sequential' => 1,
      'options' => array('sort' => 'id ASC'),
    ));

File: tests/phpunit/api/v3/JobTest.php [hunk 1]
```diff
@@ -320,16 +320,124 @@ public function testBatchMergeWorks($dataSet) {
...
+   * Group 1  Added  Added  Added
+   * Group 2  Added  Removed  ****  Added
+   * Group 3  Removed  null  **** null
+   * Group 4  Removed  Added  **** Added
+   * Group 5  Removed  Removed **** null
+   * Group 6  null  Added  Added
+   * Group 7  null  Removed  **** null
+   *
+   * The ones with **** are the ones where I think a case could be made to change the behaviour.
+   */
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/JobTest.php]: @JKingsnorth I decided to investigate (by writing a test) how groups were being handled on merge. The table above reflects what is currently happening - in short 'Removed' information is discarded.  What do you think? It feels like data loss to me - albeit I'm not sure that I mind losing that data. Should we JIRA it?

Comment (JKingsnorth) [inline tests/phpunit/api/v3/JobTest.php]: Good thinking, definitely worth the test, and I think we should JIRA it. Although it's data loss I don't think it's critical. But we should fix it up.

**Comment by eileenmcnaughton** [comment]
test this please

File: tests/phpunit/api/v3/JobTest.php (hunk 1 above)
Comment (JKingsnorth) [inline tests/phpunit/api/v3/JobTest.php]: Should be fixed by: https://github.com/civicrm/civicrm-core/pull/8702

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by JKingsnorth** [comment]
I just ran the JobTest class and I didn't see any failures like the ones @jitendrapurohit was coming up against. Jenkins test this please.

**Comment by JKingsnorth** [comment]
OK, confused now as to why my test environment is passing, but Jenkins is failing. I haven't gone through the code line-by-line, but since it is 'test-only' I expect it's OK to 'merge on pass'

**Comment by eileenmcnaughton** [comment]
incorporated into #8695

**Pull request closed without merging**